klippy/toolhead.py

```diff
@@ -501,10 +508,38 @@ def wait_moves(self):
                 break
             eventtime = self.reactor.pause(eventtime + 0.100)
     def set_extruder(self, extruder, extrude_pos):
-        self.extruder = extruder
+        # XXX - should use add_extra_axis
+        prev_ea_trapq = self.extra_axes[0].get_trapq()
+        if prev_ea_trapq in self.flush_trapqs:
+            self.flush_trapqs.remove(prev_ea_trapq)
+        self.extra_axes[0] = extruder
         self.commanded_pos[3] = extrude_pos
+        ea_trapq = extruder.get_trapq()
+        if ea_trapq is not None:
+            self.flush_trapqs.append(ea_trapq)
     def get_extruder(self):
-        return self.extruder
+        return self.extra_axes[0]
+    def add_extra_axis(self, ea, axis_pos):
+        self._flush_lookahead()
+        self.extra_axes.append(ea)
+        self.commanded_pos.append(axis_pos)
+        ea_trapq = ea.get_trapq()
+        if ea_trapq is not None:
+            self.flush_trapqs.append(ea_trapq)
+        self.printer.send_event("toolhead:update_extra_axes")
+    def remove_extra_axis(self, ea):
+        self._flush_lookahead()
+        if ea not in self.extra_axes:
+            return
+        ea_index = self.extra_axes.index(ea) + 3
+        ea_trapq = ea.get_trapq()
+        if ea_trapq in self.flush_trapqs:
+            self.flush_trapqs.remove(ea_trapq)
+        self.commanded_pos.pop(ea_index)
+        self.extra_axes.pop(ea_index - 3)
+        self.printer.send_event("toolhead:update_extra_axes")
+    def get_extra_axes(self):
+        return [None, None, None] + self.extra_axes
```

Comment:
I wonder if there is a real value in returning Nones here. Also, in this file I see a lot of instances of code fiddling with extra axes indices. Thus, would rewriting this function as
```
for i, ea in enumerate(self.extra_axes):
  yield i+3, ea
```
and calling it appropriately simplify code in toolhead.py and other files? Something to consider, perhaps.

Reply:
Yeah, I agree the indexes in toolhead.py and outside of toolhead.py are confusing.  Currently, in toolhead.py `e_index` represents the index starting at position 3, and `ea_index` represents the index starting at position 0.  Outside of toolhead.py all indexes are relative to 0.

I guess we could replace the internal toolhead `extra_axes` list with a dictionary.  It's an internal abstraction though, so not sure it matters that much.

Thanks again,
-Kevin